server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1089,17 +1186,30 @@ protected void doRun() throws Exception {
                             final Set<String> blobs = shardContainer.listBlobs().keySet();
                             final BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots;
                             final long newGen;
-                            if (useUUIDs) {
-                                newGen = -1L;
-                                blobStoreIndexShardSnapshots = buildBlobStoreIndexShardSnapshots(
-                                    blobs,
-                                    shardContainer,
-                                    oldRepositoryData.shardGenerations().getShardGen(indexId, finalShardId)
-                                ).v1();
+
+                            // Index-N file would be present if snapshots other than shallow snapshots are present for this shard
+                            if (blobs.stream()
+                                .filter(blob -> blob.startsWith(SNAPSHOT_INDEX_PREFIX))
+                                .collect(Collectors.toSet())
+                                .size() > 0) {
+                                if (useUUIDs) {
+                                    newGen = -1L;
+                                    blobStoreIndexShardSnapshots = buildBlobStoreIndexShardSnapshots(
+                                        blobs,
+                                        shardContainer,
+                                        oldRepositoryData.shardGenerations().getShardGen(indexId, finalShardId)
+                                    ).v1();
+                                } else {
+                                    Tuple<BlobStoreIndexShardSnapshots, Long> tuple = buildBlobStoreIndexShardSnapshots(
+                                        blobs,
+                                        shardContainer
+                                    );
+                                    newGen = tuple.v2() + 1;
+                                    blobStoreIndexShardSnapshots = tuple.v1();
+                                }
                             } else {
```

Comment:
would like to understand these changes better .

Reply:
Discussed offline